### PR TITLE
[soundcloud] Fix download URL with private tracks

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code (*modification*) and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Bug fix

---

### Description of your *pull request* and other information

The extractor was missing secret_token when fetching download_url, which made it invalid and yielding 404. This changeset includes some level of refactor to avoid duplication.

The "download" format name was changed into "download_url" because the error message was too confusing when it failed:
```
[soundcloud] 340344461: Checking download video format URL
[soundcloud] 340344461: download video format URL is invalid, skipping
```

A link is also provided for testing: https://soundcloud.com/oriuplift/uponly-238-no-talking-wav/s-AyZUd
I have tested with this track locally.

### CODE REVIEW

File: youtube_dl/extractor/soundcloud.py [hunk 1]
```diff
@@ -160,13 +160,16 @@ def _extract_info_dict(self, info, full_title=None, quiet=False, secret_token=No
...
         if info.get('downloadable', False):
             # We can build a direct link to the song
             format_url = (
-                'https://api.soundcloud.com/tracks/{0}/download?client_id={1}'.format(
-                    track_id, self._CLIENT_ID))
+                'https://api.soundcloud.com/tracks/{0}/download?{1}'.format(
+                    track_id, compat_urllib_parse_urlencode(query)))
             formats.append({
-                'format_id': 'download',
+                'format_id': 'download_url',
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: Revert.
>Checking download_**url** video format **URL**

makes even less sense.

File: youtube_dl/extractor/soundcloud.py [hunk 2]
```diff
@@ -160,13 +160,16 @@ def _extract_info_dict(self, info, full_title=None, quiet=False, secret_token=No
...
+        query = {'client_id': self._CLIENT_ID}
+        if secret_token is not None:
+            query['secret_token'] = secret_token
         if info.get('downloadable', False):
             # We can build a direct link to the song
             format_url = (
-                'https://api.soundcloud.com/tracks/{0}/download?client_id={1}'.format(
-                    track_id, self._CLIENT_ID))
+                'https://api.soundcloud.com/tracks/{0}/download?{1}'.format(
+                    track_id, compat_urllib_parse_urlencode(query)))
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: `update_url_query`.

File: youtube_dl/extractor/soundcloud.py [hunk 3]
```diff
@@ -216,7 +216,7 @@ def _extract_info_dict(self, info, full_title=None, quiet=False, secret_token=No
             # cannot be always used, sometimes it can give an HTTP 404 error
             formats.append({
                 'format_id': 'fallback',
-                'url': info['stream_url'] + '?client_id=' + self._CLIENT_ID,
+                'url': info['stream_url'] + compat_urllib_parse_urlencode(query),
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: `update_url_query`.

File: youtube_dl/extractor/soundcloud.py (hunk 2 above)
Comment (ishitatsuyuki) [reply youtube_dl/extractor/soundcloud.py]: The code would be longer as it is using another format parameter. It's nothing more than more overhead here.

Comment (dstftw) [reply youtube_dl/extractor/soundcloud.py]: It won't be longer.

Comment (ishitatsuyuki) [reply youtube_dl/extractor/soundcloud.py]: This is just an issue of preference. There are many other code written in this style, and on the other hand update_url_query has many redundant decode and encode.

Comment (dstftw) [reply youtube_dl/extractor/soundcloud.py]: This the matter of **code reusage**, not personal preference. You should use the most generic solution available and not **reinvent the wheel**. Other code is this style was written before `update_url_query` was introduced.

File: youtube_dl/extractor/soundcloud.py [hunk 4]
```diff
@@ -160,11 +160,13 @@ def _extract_info_dict(self, info, full_title=None, quiet=False, secret_token=No
...
+        query = {'client_id': self._CLIENT_ID}
+        if secret_token is not None:
+            query['secret_token'] = secret_token
         if info.get('downloadable', False):
             # We can build a direct link to the song
-            format_url = (
-                'https://api.soundcloud.com/tracks/{0}/download?client_id={1}'.format(
-                    track_id, self._CLIENT_ID))
+            format_url = update_url_query(
+                'https://api.soundcloud.com/tracks/{}/download'.format(track_id), query)
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: `{}` won't work in python 2.6.

**Pull request closed without merging**